From 308ce7bc0c65c600da28231a3f47e10c35c37a05 Mon Sep 17 00:00:00 2001
From: Emmanuel Berthier <emmanuel.berthier@intel.com>
Date: Thu, 7 Apr 2016 19:29:34 +0200
Subject: [PATCH 28/41] [I2C] Correct access out-of-bounds on stack in
 i2c_dw_xfer_msg

KASAN Reports:

[    3.704761] BUG: AddressSanitizer: out-of-bounds on stack in i2c_dw_xfer_msg+0x62/0x5a0
               at addr ffff88006458f1e0
[    3.716141] Read of size 2 by task swapper/0/1
[    3.721117] page:ffffea00019163c0 count:0 mapcount:0 mapping:          (null) index:0x0
[    3.730068] page flags: 0x4000000000000000()
[    3.734846] page dumped because: kasan: bad access detected
[    3.741085] CPU: 0 PID: 1 Comm: swapper/0 Not tainted 3.14.55-x86_64-00983-g396cea9-dirty #5
[    3.750524] Hardware name: Insyde CherryTrail/T3 MRD, BIOS CHTMRD.A6.001.007 03/23/2016
[    3.759576]  [<ffffffff820304fb>] dump_stack+0x4e/0x7a
[    3.765336]  [<ffffffff8127a148>] kasan_report_error+0x448/0x450
[    3.772069]  [<ffffffff81278af9>] __asan_load2+0x89/0xc0
[    3.778023]  [<ffffffff81b6b1e2>] ? i2c_dw_xfer_msg+0x62/0x5a0
[    3.784558]  [<ffffffff81b6b1e2>] i2c_dw_xfer_msg+0x62/0x5a0

Change-Id: Ia974364fa1645c2a8d7c11f98b39f6091c1625c0
Tracked-On: https://jira01.devtools.intel.com/browse/OAM-31683
Signed-off-by: Emmanuel Berthier <emmanuel.berthier@intel.com>
Reviewed-on: https://android.intel.com:443/492050
Reviewed-on: https://android.intel.com:443/504766
---
 drivers/i2c/busses/i2c-designware-core.c | 5 ++++-
 1 file changed, 4 insertions(+), 1 deletion(-)

diff --git a/drivers/i2c/busses/i2c-designware-core.c b/drivers/i2c/busses/i2c-designware-core.c
index 4b9fcfa..eb4a743e 100644
--- a/drivers/i2c/busses/i2c-designware-core.c
+++ b/drivers/i2c/busses/i2c-designware-core.c
@@ -497,11 +497,14 @@ i2c_dw_xfer_msg(struct dw_i2c_dev *dev)
 	struct i2c_msg *msgs = dev->msgs;
 	u32 intr_mask;
 	int tx_limit, rx_limit;
-	u32 addr = msgs[dev->msg_write_idx].addr;
+	u32 addr = 0;
 	u32 buf_len = dev->tx_buf_len;
 	u8 *buf = dev->tx_buf;
 	bool need_restart = false;
 
+	if (dev->msg_write_idx < dev->msgs_num)
+		addr = msgs[dev->msg_write_idx].addr;
+
 	intr_mask = DW_IC_INTR_DEFAULT_MASK;
 
 	for (; dev->msg_write_idx < dev->msgs_num; dev->msg_write_idx++) {
-- 
1.9.1

